utils/src/set.rs

```diff
@@ -457,6 +485,229 @@ impl<K, V> DoubleEndedIterator for OrderedAssociatedIntoIter<K, V> {
     }
 }
 
+/// An ordered, deduplicated slice of items each paired with some associated value, where values must be unique.
+///
+/// Like [`OrderedAssociated`], but enforces that values are unique across all keys. The contained
+/// [`Vec<(K, V)>`] is sealed after construction and cannot be modified. To unseal the inner
+/// [`Vec<(K, V)>`], use the [`Into<Vec<(K, V)>>`] impl.
+///
+/// Consumers that only need the ordered keys can treat an [`OrderedInjection`] as an
+/// [`Ordered`] through deref coercions.
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct OrderedInjection<K, V> {
+    inner: OrderedAssociated<K, V>,
+}
+
+impl<K, V> OrderedInjection<K, V> {
+    /// Returns the number of entries in the map.
+    pub const fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns `true` if the map is empty.
+    pub const fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Returns a key by index, if it exists.
+    pub fn get(&self, index: usize) -> Option<&K> {
+        self.inner.get(index)
+    }
+
+    /// Returns the position of the provided key, if it exists.
+    pub fn position(&self, key: &K) -> Option<usize>
+    where
+        K: Ord,
+    {
+        self.inner.position(key)
+    }
+
+    /// Returns the ordered keys as an [`Ordered`] reference.
+    pub const fn keys(&self) -> &Ordered<K> {
+        self.inner.keys()
+    }
+
+    /// Consumes the map and returns the ordered keys.
+    pub fn into_keys(self) -> Ordered<K> {
+        self.inner.into_keys()
+    }
+
+    /// Returns the associated value at `index`, if it exists.
+    pub fn value(&self, index: usize) -> Option<&V> {
+        self.inner.value(index)
+    }
+
+    /// Returns the associated value for `key`, if it exists.
+    pub fn get_value(&self, key: &K) -> Option<&V>
+    where
+        K: Ord,
+    {
+        self.inner.get_value(key)
+    }
+
+    /// Returns the associated values.
+    pub fn values(&self) -> &[V] {
+        self.inner.values()
+    }
+
+    /// Returns a zipped iterator over keys and values.
+    pub fn iter_pairs(&self) -> impl Iterator<Item = (&K, &V)> {
+        self.inner.iter_pairs()
+    }
+
+    /// Returns an iterator over the ordered keys.
+    pub fn iter(&self) -> core::slice::Iter<'_, K> {
+        self.inner.iter()
+    }
+
+    /// Attempts to create an [`OrderedInjection`] from an iterator of key-value pairs.
+    ///
+    /// Returns an error if any value is duplicated across different keys.
+    pub fn try_from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Result<Self, Error>
+    where
+        K: Ord,
+        V: Eq,
+    {
+        let map: OrderedAssociated<K, V> = iter.into_iter().collect();
+        Self::try_from(map)
+    }
+}
+
+impl<K, V: Eq> TryFrom<OrderedAssociated<K, V>> for OrderedInjection<K, V> {
```

Comment:
If we made `V: Eq + Hash`, we could use a HashSet here (making this `O(n)` rather than `O(n^2)`.